hrms/api/roster.py

```diff
@@ -0,0 +1,258 @@
+import frappe
+from frappe import _
+from frappe.utils import add_days, date_diff, random_string
+
+from hrms.hr.doctype.shift_assignment.shift_assignment import ShiftAssignment
+from hrms.hr.doctype.shift_assignment_tool.shift_assignment_tool import create_shift_assignment
+
+
+@frappe.whitelist()
+def get_values(doctype: str, name: str, fields: list) -> dict[str, str]:
+	return frappe.db.get_value(doctype, name, fields, as_dict=True)
+
+
+@frappe.whitelist()
+def get_events(
+	month_start: str, month_end: str, employee_filters: dict[str, str], shift_filters: dict[str, str]
+) -> dict[str, list[dict]]:
+	holidays = get_holidays(month_start, month_end, employee_filters)
+	leaves = get_leaves(month_start, month_end, employee_filters)
+	shifts = get_shifts(month_start, month_end, employee_filters, shift_filters)
+
+	events = {}
+	for event in [holidays, leaves, shifts]:
+		event = group_by_employee(event)
+		for key, value in event.items():
+			if key in events:
+				events[key].extend(value)
+			else:
+				events[key] = value
+	return events
+
+
+@frappe.whitelist()
+def create_shift_assignment_schedule(
+	employee: str,
+	company: str,
+	shift_type: str,
+	status: str,
+	start_date: str,
+	end_date: str | None,
+	days: list[str],
+	frequency: str,
+) -> None:
+	schedule = frappe.get_doc(
+		{
+			"doctype": "Shift Assignment Schedule",
+			"schedule": random_string(10),
+			"frequency": frequency,
+			"days": [{"day": day} for day in days],
+			"status": "Inactive" if end_date else "Active",
+			"employee": employee,
+			"company": company,
+			"shift_type": shift_type,
+			"shift_status": status,
+		}
+	).insert()
+
+	if not end_date or date_diff(end_date, start_date) <= 90:
+		return schedule.create_shifts(start_date, end_date)
+
+	frappe.enqueue(schedule.create_shifts, timeout=4500, start_date=start_date, end_date=end_date)
+
+
+@frappe.whitelist()
+def delete_shift_assignment_schedule(schedule: str) -> None:
+	frappe.db.delete("Shift Assignment", {"schedule": schedule})
+	frappe.delete_doc("Shift Assignment Schedule", schedule)
+
+
+@frappe.whitelist()
+def swap_shift(
```

Comment:
Here I pushed Day Shift on Sun 02 from Rucha -> Ramesh. Now I am trying to put it back to Rucha's row but this message shows up. It should allow this right?
![image](https://github.com/frappe/hrms/assets/24353136/b0086ad3-8d6a-44ee-b77d-e62c0af72c7b)

Also, I might accidentally swap shifts, it should warn me before doing so. We need to figure out better UX for swapping shifts. I think for now this can happen via a modal instead of direct drag and drop with explicit employee selections and days. Not just swapping but reassigning too.
A dialog would just gives more control and context on what's happening.

Or we could leave it out in the first cut. What do you think?